lib/web/resources.go

```diff
@@ -293,23 +302,65 @@ func upsertTrustedCluster(ctx context.Context, clt resourcesAPIGetter, content,
 	return ui.NewResourceItem(tc)
 }
 
-// CheckResourceUpsertableByError checks if the resource is upsertable by the state of error with
-// the request http method used.
-func CheckResourceUpsertableByError(err error, httpMethod, resourceName string) error {
-	if err != nil && !trace.IsNotFound(err) {
-		return trace.Wrap(err)
+// getResource tries to retrieve a resource (by name),
+// returning a NotFound error if the resource does not exist.
+type getResource func(context.Context, string) (types.Resource, error)
+
+// CheckResourceUpsert checks if the resource can be created or updated, depending on the http method.
+func CheckResourceUpsert(ctx context.Context, httpMethod string, params httprouter.Params, payloadResourceName string, get getResource) error {
+	switch httpMethod {
+	case http.MethodPost:
+		return trace.Wrap(checkResourceCreate(ctx, payloadResourceName, get))
+	case http.MethodPut:
+		resourceName := params.ByName("name")
+		if resourceName == "" {
+			return trace.BadParameter("missing resource name")
+		}
+		return trace.Wrap(checkResourceUpdate(ctx, payloadResourceName, resourceName, get))
+	default:
+		return trace.NotImplemented("http method %q not expected. this is a bug!", httpMethod)
+	}
+}
+
+// checkResourceCreate checks if the resource can be created, returning nil if it can.
+func checkResourceCreate(ctx context.Context, payloadResourceName string, get getResource) error {
+	// Try to retrieve the resource by name.
+	_, err := get(ctx, payloadResourceName)
```

Comment:
why do you need an extra get here, wouldn't Create return an AlreadyExists error?

Reply:
Could you please clarify what you mean by an extra get?

Create returns:
- `AlreadyExists` if the `get` does not return an error (which happens if a resource with the same name already exists)
- `nil` if the `get` returns `NotFound` (which happens if no resource with the same name exists)
- some other error otherwise

Reply:
Right. My thought was that it does a `Get(Type)` to check if it already exists, but if you just called `Create(Type)` it would be an error anyways if the resource already exists. So the `Get(Type)` is technically unecessary.

But after a re-read this morning, it looks like the old code did the same thing, and actually it doesn't ever call `Create(Type)` but uses `Upsert(Type)` for both PUT and POST, so that wouldn't work. It seems okay, then